File: frontend/src/utils/ean13.js
export const AUTO_EAN13_SENTINEL = "__AUTO_EAN13__"

export function normalizeBarcodeDigits(value) {
  return String(value || "").replace(/\D+/g, "")
}

export function isValidEan13(value) {
  const digits = normalizeBarcodeDigits(value)
  if (!/^\d{13}$/.test(digits)) return false
  return computeEan13CheckDigit(digits.slice(0, 12)) === digits.slice(12)
}

export function computeEan13CheckDigit(body12) {
  const digits = normalizeBarcodeDigits(body12)
  if (!/^\d{12}$/.test(digits)) throw new Error("EAN-13 body must have 12 digits")
  let sum = 0
  for (let i = 0; i < digits.length; i += 1) {
    const n = Number(digits[i])
    sum += i % 2 === 0 ? n : n * 3
  }
  return String((10 - (sum % 10)) % 10)
}

export function buildEan13(body12) {
  const digits = normalizeBarcodeDigits(body12)
  if (!/^\d{12}$/.test(digits)) throw new Error("EAN-13 body must have 12 digits")
  return `${digits}${computeEan13CheckDigit(digits)}`
}

export function barcodeBcidForValue(value) {
  return isValidEan13(value) ? "ean13" : "code128"
}

File: backend/app/services/ean13.py
from __future__ import annotations

from secrets import randbelow


AUTO_EAN13_SENTINEL = "__AUTO_EAN13__"
DEFAULT_EAN13_PREFIX = "21"


def normalize_barcode_digits(value: str | None) -> str:
    return "".join(ch for ch in str(value or "") if ch.isdigit())


def compute_ean13_check_digit(body12: str) -> str:
    digits = normalize_barcode_digits(body12)
    if len(digits) != 12:
        raise ValueError("EAN-13 body must have 12 digits")
    total = 0
    for idx, ch in enumerate(digits):
        n = int(ch)
        total += n if idx % 2 == 0 else n * 3
    return str((10 - (total % 10)) % 10)


def build_ean13(body12: str) -> str:
    digits = normalize_barcode_digits(body12)
    if len(digits) != 12:
        raise ValueError("EAN-13 body must have 12 digits")
    return f"{digits}{compute_ean13_check_digit(digits)}"


def is_valid_ean13(value: str | None) -> bool:
    digits = normalize_barcode_digits(value)
    return len(digits) == 13 and compute_ean13_check_digit(digits[:12]) == digits[12]


def generate_random_ean13(prefix: str = DEFAULT_EAN13_PREFIX) -> str:
    prefix_digits = normalize_barcode_digits(prefix or DEFAULT_EAN13_PREFIX) or DEFAULT_EAN13_PREFIX
    if len(prefix_digits) >= 12:
        body12 = prefix_digits[:12]
    else:
        remaining = 12 - len(prefix_digits)
        suffix = "".join(str(randbelow(10)) for _ in range(remaining))
        body12 = f"{prefix_digits}{suffix}"
    return build_ean13(body12)
